File: PCA/amplapi/examples/csharp/OptionsExample.cs
using ampl;
using System;

namespace Examples
{
  public class OptionsExample
  {
    public static int Main(string[] args)
    {
      /*
      // If the AMPL installation directory is not in the system search path:
      ampl.Environment env = new ampl.Environment(
        "full path to the AMPL installation directory");
      // Create an AMPL instance
      using (AMPL a = new AMPL(env)) {}
      */

      // Create an AMPL instance
      using (AMPL ampl = new AMPL())
      {
        // Get the value of the option preSolve and print
        int presolve = ampl.GetIntOption("presolve").Value;
        Console.Write("AMPL preSolve is {0}\n", presolve);

        // set the value to false (maps to 0)
        ampl.SetOption("presolve", false);

        // Get the value of the option preSolve and print
        presolve = ampl.GetIntOption("presolve").Value;
        Console.Write("AMPL preSolve is now {0}\n", presolve);

        // Check whether an option with a specified name exists
        string value = ampl.GetOption("Solver");
        if (value != null)
          Console.WriteLine("Option Solver exists and has value: " + value);

        // Check again, this time failing
        value = ampl.GetOption("s_o_l_v_e_r");
        if (value == null)
          Console.WriteLine("Option s_o_l_v_e_r does not exists!");

        // Using nullables (for bool, int and double options)
        double? dblvalue = ampl.GetDblOption("does_not_exist");
        if (dblvalue.HasValue)
          Console.WriteLine("Surprisingly, \"does_not_exist\" has a value!");
        else
          Console.WriteLine("\"does_not_exist\" does not have a value!");

        // Error if accessing value
        try
        {
          Console.WriteLine(dblvalue.Value);
        }
        catch (InvalidOperationException ex)
        {
          Console.WriteLine("Error while accessing not existing value:");
          Console.WriteLine(ex.Message);
        }
      }
      return 0;
    }
  }
}

File: PCA/amplapi/examples/cpp/optionsexample.cc
#include "ampl/ampl.h"

#include <iostream>

int main() {
  try {
    // Create an AMPL instance
    ampl::AMPL ampl;

    /*
    // If the AMPL installation directory is not in the system search path:
    ampl::Environment env("full path to the AMPL installation directory");
    ampl::AMPL ampl(env);
    */

    // Get the value of the option presolve and print
    int presolve = *ampl.getIntOption("presolve");
    int s;
#ifdef FMT_GCC_VERSION
    s = FMT_GCC_VERSION;
#else
    s = -1;
#endif
    std::cout << "AMPL presolve is " << presolve << "\n";
    std::cout << "FMT_GCC_VERSION = " << s << "\n";
    // Set the value to false (maps to 0)
    ampl.setBoolOption("presolve", false);

    // Get the value of the option presolve and print
    presolve = *ampl.getIntOption("presolve");
    std::cout << "AMPL presolve is now " << presolve << "\n";

    // Check whether an option with a specified name
    // exists
    if (ampl::Optional<std::string> value =
      ampl.getOption("solver")) {
      std::cout << "Option solver exists and has value: "
        << *value << "\n";
    }
    // Check again, this time failing
    if (ampl::Optional<std::string> value =
      ampl.getOption("s_o_l_v_e_r")) {
      std::cout << "Option s_o_l_v_e_r does not exists!\n";
    }
    // Optional::value() throws exception if it has no value
    try {
      ampl.getOption("s_o_l_v_e_r").value();
    }
    catch (const std::exception &ex) {
      std::cout << "Option s_o_l_v_e_r does not exist" << std::endl;
      std::cout << ex.what() << std::endl;
    }
    return 0;
  } catch (const std::exception &e) {
    std::cout << e.what() << "\n";
    return 1;
  }
}
